In [2]:
from tensorflow.keras.models import *
from tensorflow.keras.layers import *
from tensorflow.keras import Input
from tensorflow.keras import Sequential
from tensorflow.keras.metrics import *
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder # iris dataset
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd

# Linear Regression

# First Dataset

In [423]:
data = pd.read_csv("ps1data1.txt")

In [424]:
X = data.iloc[ : , 0:len(data.columns)-1]
Y = data.iloc[ : , len(data.columns)-1:len(data.columns)]

In [425]:
scalex = MinMaxScaler()
scaley = MinMaxScaler()

In [426]:
X = scalex.fit_transform(X)
Y = scaley.fit_transform(Y)

In [427]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=123)

In [428]:
model = Sequential()

model.add(Input(shape=(10,), dtype="float32") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(1, activation="linear") )

model.summary()

Model: "sequential_22"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_70 (Dense)             (None, 32)                352       
_________________________________________________________________
dense_71 (Dense)             (None, 32)                1056      
_________________________________________________________________
dense_72 (Dense)             (None, 1)                 33        
Total params: 1,441
Trainable params: 1,441
Non-trainable params: 0
_________________________________________________________________


In [429]:
model.compile( optimizer="adam", loss="mse")

In [430]:
model.fit( x_train, y_train, epochs=10, batch_size=128, validation_split=0.1 )

Epoch 1/10
563/563 [==============================] - 2s 2ms/step - loss: 0.0093 - val_loss: 3.9674e-05
Epoch 2/10
563/563 [==============================] - 1s 2ms/step - loss: 3.8504e-05 - val_loss: 3.6837e-05
Epoch 3/10
563/563 [==============================] - 1s 2ms/step - loss: 3.6308e-05 - val_loss: 3.8940e-05
Epoch 4/10
563/563 [==============================] - 1s 2ms/step - loss: 3.4423e-05 - val_loss: 4.3900e-05
Epoch 5/10
563/563 [==============================] - 1s 2ms/step - loss: 3.2711e-05 - val_loss: 3.0601e-05
Epoch 6/10
563/563 [==============================] - 1s 2ms/step - loss: 3.2570e-05 - val_loss: 2.7749e-05
Epoch 7/10
563/563 [==============================] - 1s 2ms/step - loss: 2.9760e-05 - val_loss: 2.8453e-05
Epoch 8/10
563/563 [==============================] - 1s 2ms/step - loss: 2.9912e-05 - val_loss: 2.5638e-05
Epoch 9/10
563/563 [==============================] - 1s 2ms/step - loss: 2.8113e-05 - val_loss: 2.8662e-05
Epoch 10/10
563/563 [===========

In [431]:
model.evaluate(x_test, y_test)

625/625 [==============================] - 1s 2ms/step - loss: 2.2742e-05


2.2742406144971028e-05

# Second Dataset

In [411]:
data = pd.read_csv("ps1data2.txt")

In [412]:
X = data.iloc[ : , 0:len(data.columns)-1]
Y = data.iloc[ : , len(data.columns)-1:len(data.columns)]

In [413]:
scalex = MinMaxScaler()
scaley = MinMaxScaler()

In [414]:
X = scalex.fit_transform(X)
Y = scaley.fit_transform(Y)

In [415]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=123)

In [416]:
model = Sequential()

model.add(Input(shape=(10,), dtype="float32") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(1, activation="linear") )

model.summary()

Model: "sequential_21"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_67 (Dense)             (None, 32)                352       
_________________________________________________________________
dense_68 (Dense)             (None, 32)                1056      
_________________________________________________________________
dense_69 (Dense)             (None, 1)                 33        
Total params: 1,441
Trainable params: 1,441
Non-trainable params: 0
_________________________________________________________________


In [417]:
model.compile( optimizer="adam", loss="mse")

In [418]:
model.fit( x_train, y_train, epochs=10, batch_size=128, validation_split=0.1 )

Epoch 1/10
563/563 [==============================] - 2s 2ms/step - loss: 0.0042 - val_loss: 0.0025
Epoch 2/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0024 - val_loss: 0.0023
Epoch 3/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0024 - val_loss: 0.0023
Epoch 4/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0024
Epoch 5/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0024 - val_loss: 0.0023
Epoch 6/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0023
Epoch 7/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0022
Epoch 8/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0023
Epoch 9/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0023
Epoch 10/10
563/563 [==============================] - 1s 2ms/step - loss: 0.0023 - val_loss: 0.0021

In [419]:
model.evaluate(x_test, y_test)

625/625 [==============================] - 1s 2ms/step - loss: 0.0022


0.002228186698630452

## Prediction

In [420]:
prediction = model.predict(x_test.iloc[0:1])

Scaled down:

In [421]:
print(prediction[0][0])

-0.048002943


Scaled up to original:

In [422]:
print(scaley.inverse_transform(prediction)[0][0])

-2189280.8


# Third Dataset

In [3]:
data = pd.read_csv("ps1data3.txt")
data

509.0   576.0   546.0   3974.0   2706.0   3595.0   383.0   1129.0  \
0      877.0  1810.0   491.0   3074.0   3945.0   3836.0  2540.0   1853.0   
1     1987.0  1753.0   468.0   3337.0   2668.0   1266.0    15.0   2819.0   
2     1849.0  2813.0   391.0   1755.0   2417.0   1180.0  1402.0   3239.0   
3      239.0  1854.0   138.0   1984.0   3824.0    102.0  2798.0   1548.0   
4      766.0  1555.0   458.0   3220.0    216.0   4744.0     7.0   3505.0   
...      ...     ...     ...      ...      ...      ...     ...      ...   
9994  2098.0  3575.0   203.0    132.0   2705.0   3984.0  2654.0   1597.0   
9995  1669.0  3852.0    42.0   4427.0   1014.0   3063.0  1657.0   3434.0   
9996   824.0  3664.0   483.0     50.0    451.0    785.0  2485.0   1551.0   
9997  1857.0  2761.0     6.0   3545.0   3691.0   2623.0  2589.0   2998.0   
9998   970.0  1752.0   522.0    646.0    570.0   3570.0  2532.0   3371.0   

       532.0   546.0.1  ...   19.0.4   2126.0   3304.0   114.0   621.0  \
0     1009.0     888.0  ...     18.0   1643.0   4271.0  1089.0   938.0   
1     1455.0     826.0  ...     41.0   3624.0    420.0   106.0   487.0   
2      251.0     448.0  ...      9.0   2809.0   1612.0   188.0  1739.0   
3       23.0     808.0  ...     45.0   2371.0   2618.0   706.0   609.0   
4     1474.0    1435.0  ...     39.0    145.0    559.0    36.0  1465.0   
...      ...       ...  ...      ...      ...      ...     ...     ...   
9994  1645.0     786.0  ...     60.0   3892.0   2109.0  1216.0  1055.0   
9995   886.0     318.0  ...     49.0    685.0   1497.0  1455.0  1672.0   
9996   768.0     344.0  ...     53.0   2533.0   1517.0   880.0  2577.0   
9997   891.0    1473.0  ...     50.0   1114.0   3192.0   321.0  1468.0   
9998  1195.0     345.0  ...     47.0   3205.0   4119.0  1461.0  1303.0   

       1048.0   146.0   16.0.1   239.0   324.03  
0       480.0   302.0      4.0   906.0  2085.70  
1       182.0   472.0     75.0  2073.0  2833.40  
2      1991.0   338.0     75.0  1154.0  1385.15  
3       437.0   375.0    155.0    72.0  2611.66  
4      1092.0    23.0      4.0  2302.0  2285.97  
...       ...     ...      ...     ...      ...  
9994   1076.0    82.0      2.0   509.0   800.42  
9995   1573.0   342.0    174.0   675.0  2711.37  
9996      3.0   175.0    121.0  1088.0  1354.40  
9997   1916.0   105.0      1.0   751.0  3676.06  
9998    632.0   123.0    161.0  1584.0  1706.72  

[9999 rows x 1001 columns]

In [4]:
X = data.iloc[ : , 0:len(data.columns)-1]
Y = data.iloc[ : , len(data.columns)-1:len(data.columns)]

In [5]:
scalex = MinMaxScaler()
scaley = MinMaxScaler()

In [6]:
X = scalex.fit_transform(X)
Y = scaley.fit_transform(Y)

In [7]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=123)

In [8]:
model = Sequential()

model.add(Input(shape=(1000,), dtype="float32") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(1, activation="linear") )

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 32)                32032     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 33        
Total params: 32,065
Trainable params: 32,065
Non-trainable params: 0
_________________________________________________________________


In [9]:
model.compile( optimizer="adam", loss="mse")

In [10]:
model.fit( x_train, y_train, epochs=100, batch_size=128, validation_split=0.1 )

Epoch 1/100
57/57 [==============================] - 1s 5ms/step - loss: 0.0551 - val_loss: 0.0325
Epoch 2/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0311 - val_loss: 0.0292
Epoch 3/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0282 - val_loss: 0.0277
Epoch 4/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0269 - val_loss: 0.0266
Epoch 5/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0259 - val_loss: 0.0273
Epoch 6/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0244 - val_loss: 0.0250
Epoch 7/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0228 - val_loss: 0.0226
Epoch 8/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0211 - val_loss: 0.0214
Epoch 9/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0197 - val_loss: 0.0224
Epoch 10/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0190 - val_loss: 0.0235
Epoch 11/

57/57 [==============================] - 0s 2ms/step - loss: 0.0115 - val_loss: 0.0174
Epoch 84/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0107 - val_loss: 0.0173
Epoch 85/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0105 - val_loss: 0.0174
Epoch 86/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0102 - val_loss: 0.0174
Epoch 87/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0106 - val_loss: 0.0175
Epoch 88/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0101 - val_loss: 0.0203
Epoch 89/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0101 - val_loss: 0.0183
Epoch 90/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0102 - val_loss: 0.0174
Epoch 91/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0095 - val_loss: 0.0189
Epoch 92/100
57/57 [==============================] - 0s 2ms/step - loss: 0.0095 - val_loss: 0.0186
Epoch 93/100


In [11]:
model.evaluate(x_test, y_test)

63/63 [==============================] - 0s 1ms/step - loss: 0.0207


0.020743288099765778

In [19]:
weights = model.get_weights()[0]
length = len(weights)
for i in range(0, length):
    if(sum(abs(weights[i][:])) > 2):
        print(sum(abs(weights[i])),i+1)

2.07029939617496 41
2.03844708824181 63
2.153975899098441 82
11.706265948712826 101
12.860410548746586 129
2.192574010696262 148
2.0009046476334333 151
2.0274893956957385 222
2.0078923739492893 225
2.31528353667818 476
2.129663037834689 499
2.0330728499684483 720
2.023197618982522 722
2.0733554986072704 764
2.0569445767905563 859
2.0585444010794163 878
2.027658084174618 926
2.0906489703047555 955
2.0239898548461497 1000


feature 101 and feature 129 had the most impact on the prediction

# Iris Dataset

In [22]:
data = pd.read_csv("irisdata.txt")

In [23]:
X = data.iloc[ : , 0:len(data.columns)-1]
Y = data.iloc[ : , len(data.columns)-1:len(data.columns)]
# for manual scaling later in get_flower_type method
columns = X.columns
maxs = [max(X[columns[0]]), max(X[columns[1]]), max(X[columns[2]]), max(X[columns[3]])]
mins = [min(X[columns[0]]),  min(X[columns[1]]), min(X[columns[2]]), min(X[columns[3]])]

In [24]:
scalex = MinMaxScaler()
encoder = OneHotEncoder(sparse=False)

In [25]:
X = scalex.fit_transform(X)
Y = encoder.fit_transform(Y)

In [26]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)
x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=123)

In [27]:
model = Sequential()

model.add(Input(shape=(4,), dtype="float32") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(32, activation="sigmoid") )
model.add( Dense(3, activation="softmax") )

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 32)                160       
_________________________________________________________________
dense_3 (Dense)              (None, 32)                1056      
_________________________________________________________________
dense_4 (Dense)              (None, 3)                 99        
Total params: 1,315
Trainable params: 1,315
Non-trainable params: 0
_________________________________________________________________


## Metrics

## Import tensorflow.keras.metrics.* to use metrics such as precision and recall

In [28]:
model.compile( optimizer="adam", loss="mse" , metrics=["accuracy", Precision(), Recall()])

In [29]:
model.fit( x_train, y_train, epochs=200, batch_size=128, validation_split=0.1 )

Epoch 1/200
1/1 [==============================] - 1s 659ms/step - loss: 0.2195 - accuracy: 0.3645 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2179 - val_accuracy: 0.4167 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 2/200
1/1 [==============================] - 0s 24ms/step - loss: 0.2191 - accuracy: 0.3645 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2202 - val_accuracy: 0.4167 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 3/200
1/1 [==============================] - 0s 24ms/step - loss: 0.2189 - accuracy: 0.3645 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2220 - val_accuracy: 0.4167 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 4/200
1/1 [==============================] - 0s 25ms/step - loss: 0.2188 - accuracy: 0.3645 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2232 - val_accuracy: 0.4167 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 5/200
1/1 [==============================

1/1 [==============================] - 0s 24ms/step - loss: 0.2119 - accuracy: 0.6916 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2180 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 36/200
1/1 [==============================] - 0s 16ms/step - loss: 0.2116 - accuracy: 0.6916 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2179 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 37/200
1/1 [==============================] - 0s 24ms/step - loss: 0.2113 - accuracy: 0.6916 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2178 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 38/200
1/1 [==============================] - 0s 17ms/step - loss: 0.2110 - accuracy: 0.6916 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2177 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 39/200
1/1 [==============================] - 0s 16

1/1 [==============================] - 0s 24ms/step - loss: 0.1962 - accuracy: 0.6636 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2106 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 70/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1956 - accuracy: 0.6636 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2102 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 71/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1949 - accuracy: 0.6636 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2098 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 72/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1942 - accuracy: 0.6636 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 0.2094 - val_accuracy: 0.5000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 73/200
1/1 [==============================] - 0s 24

1/1 [==============================] - 0s 24ms/step - loss: 0.1663 - accuracy: 0.8037 - precision: 1.0000 - recall: 0.3178 - val_loss: 0.1911 - val_accuracy: 0.7500 - val_precision: 1.0000 - val_recall: 0.0833
Epoch 105/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1653 - accuracy: 0.8131 - precision: 1.0000 - recall: 0.3271 - val_loss: 0.1904 - val_accuracy: 0.8333 - val_precision: 1.0000 - val_recall: 0.0833
Epoch 106/200
1/1 [==============================] - 0s 22ms/step - loss: 0.1643 - accuracy: 0.8224 - precision: 1.0000 - recall: 0.3271 - val_loss: 0.1896 - val_accuracy: 0.8333 - val_precision: 1.0000 - val_recall: 0.0833
Epoch 107/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1633 - accuracy: 0.8318 - precision: 1.0000 - recall: 0.3271 - val_loss: 0.1889 - val_accuracy: 0.8333 - val_precision: 1.0000 - val_recall: 0.0833
Epoch 108/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1623 - accuracy: 0.8411 - precision: 1.000

Epoch 141/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1316 - accuracy: 0.8598 - precision: 1.0000 - recall: 0.3645 - val_loss: 0.1614 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.0833
Epoch 142/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1308 - accuracy: 0.8785 - precision: 1.0000 - recall: 0.3832 - val_loss: 0.1606 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.1667
Epoch 143/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1300 - accuracy: 0.8785 - precision: 1.0000 - recall: 0.4112 - val_loss: 0.1597 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.1667
Epoch 144/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1292 - accuracy: 0.8785 - precision: 1.0000 - recall: 0.4112 - val_loss: 0.1589 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.1667
Epoch 145/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1284 - accuracy: 0.8785 - pr

1/1 [==============================] - 0s 24ms/step - loss: 0.1059 - accuracy: 0.9346 - precision: 1.0000 - recall: 0.5514 - val_loss: 0.1329 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.4167
Epoch 178/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1053 - accuracy: 0.9346 - precision: 1.0000 - recall: 0.5607 - val_loss: 0.1322 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.4167
Epoch 179/200
1/1 [==============================] - 0s 16ms/step - loss: 0.1046 - accuracy: 0.9346 - precision: 1.0000 - recall: 0.5607 - val_loss: 0.1314 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.5000
Epoch 180/200
1/1 [==============================] - 0s 25ms/step - loss: 0.1040 - accuracy: 0.9346 - precision: 1.0000 - recall: 0.5607 - val_loss: 0.1306 - val_accuracy: 1.0000 - val_precision: 1.0000 - val_recall: 0.5000
Epoch 181/200
1/1 [==============================] - 0s 24ms/step - loss: 0.1034 - accuracy: 0.9346 - precision: 1.000

In [30]:
model.evaluate(x_test, y_test)

1/1 [==============================] - 0s 16ms/step - loss: 0.0734 - accuracy: 0.9000 - precision: 1.0000 - recall: 0.8667


[0.07336597144603729, 0.8999999761581421, 1.0, 0.8666666746139526]

In [31]:
model.save('iris')

INFO:tensorflow:Assets written to: iris\assets


## Classification

In [32]:
# caveat is the data has to be loaded at some point as well so we know what the max and min is so we can scale the input
# since the weights are meant for scaled data
def get_flower_type(arg1,arg2,arg3,arg4):
#     flower labels
    labels = ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
#     load model
    localmodel = load_model('iris')
#     scale input
    inputx = [[arg1,arg2,arg3,arg4]]
    for i in range(0,3):
        inputx[0][i] = (inputx[0][i] - mins[i]) / (maxs[i] - mins[i])
    prediction = localmodel.predict(inputx)
#     find highest scoring label
    maximum = 0
    index = 0;
    for i in range(0,2):
        if(prediction[0][i] > maximum):
            maximum = prediction[0][i]
            index = i
    return labels[index]

In [33]:
get_flower_type(6.5,3.0,5.5,1.8)

'Iris-versicolor'

# Transfer Learning

In [387]:
data = pd.read_csv("ps1data1.txt")

In [388]:
X = data.iloc[ : , 0:len(data.columns)-1]
Y = data.iloc[ : , len(data.columns)-1:len(data.columns)]

In [389]:
scalex = MinMaxScaler()
scaley = MinMaxScaler()

In [390]:
X = scalex.fit_transform(X)
Y = scaley.fit_transform(Y)

In [391]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=123)

In [392]:
model = Sequential()

model.add(Input(shape=(10,), dtype="float32") )
model.add( Dense(10, activation="relu") )
model.add( Dense(10, activation="relu") )
model.add( Dense(1, activation="linear") )

model.summary()

Model: "sequential_19"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_61 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_62 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_63 (Dense)             (None, 1)                 11        
Total params: 231
Trainable params: 231
Non-trainable params: 0
_________________________________________________________________


In [393]:
model.compile( optimizer="adam", loss="mse")

In [394]:
model.fit( x_train, y_train, epochs=2, batch_size=128, validation_split=0.1 )

Epoch 1/2
563/563 [==============================] - 2s 2ms/step - loss: 0.0253 - val_loss: 0.0010
Epoch 2/2
563/563 [==============================] - 1s 2ms/step - loss: 4.4366e-04 - val_loss: 2.0065e-04


In [395]:
model.evaluate(x_test, y_test)

625/625 [==============================] - 1s 2ms/step - loss: 1.8462e-04


0.00018462225852999836

In [396]:
model.save('ps1model')

INFO:tensorflow:Assets written to: ps1model\assets


In [397]:
diskmodel = load_model('ps1model')
diskmodel.summary()

Model: "sequential_19"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_61 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_62 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_63 (Dense)             (None, 1)                 11        
Total params: 231
Trainable params: 231
Non-trainable params: 0
_________________________________________________________________


In [398]:
diskmodel.pop()
diskmodel.add( Dense(20, activation='relu') )
diskmodel.add( Dense(1, activation='linear') )
diskmodel.summary()

Model: "sequential_19"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_61 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_62 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_64 (Dense)             (None, 20)                220       
_________________________________________________________________
dense_65 (Dense)             (None, 1)                 21        
Total params: 461
Trainable params: 461
Non-trainable params: 0
_________________________________________________________________


In [399]:
diskmodel.compile( optimizer="adam", loss="mse")

In [400]:
diskmodel.fit( x_train, y_train, epochs=10, batch_size=128, validation_split=0.1 )

Epoch 1/10
563/563 [==============================] - 2s 2ms/step - loss: 0.0165 - val_loss: 1.1907e-04
Epoch 2/10
563/563 [==============================] - 1s 2ms/step - loss: 8.7736e-05 - val_loss: 7.4093e-05
Epoch 3/10
563/563 [==============================] - 1s 2ms/step - loss: 6.1058e-05 - val_loss: 5.4431e-05
Epoch 4/10
563/563 [==============================] - 1s 2ms/step - loss: 4.6723e-05 - val_loss: 4.6801e-05
Epoch 5/10
563/563 [==============================] - 1s 2ms/step - loss: 3.7509e-05 - val_loss: 3.5693e-05
Epoch 6/10
563/563 [==============================] - 1s 2ms/step - loss: 3.2008e-05 - val_loss: 3.0444e-05
Epoch 7/10
563/563 [==============================] - 1s 2ms/step - loss: 2.7985e-05 - val_loss: 2.8913e-05
Epoch 8/10
563/563 [==============================] - 1s 2ms/step - loss: 2.5637e-05 - val_loss: 2.7139e-05
Epoch 9/10
563/563 [==============================] - 1s 2ms/step - loss: 2.4893e-05 - val_loss: 2.3310e-05
Epoch 10/10
563/563 [===========